src/lib.rs

```diff
@@ -268,21 +271,95 @@ impl<'object, Endian> DebugInfo<'object, Endian>
+            let mut rows = {
+                // First, check our binary search list
+                if let Ok(skiplist) = unit.skiplist.read() {
+                    match skiplist.binary_search_by_key(&addr, |&(raddr, _, _)| raddr) {
+                        Ok(i) => {
+                            // we have a state machine for this address!
+                            current = Some(skiplist[i].2);
+                            rowi = (i + 1) * self.cache_every;
+                            skiplist[i].1.clone()
+                        }
+                        Err(i) => {
+                            // i is the entry in the skiplist where addr *would* appear. Thus, we
+                            // don't have a state machine for *this* address, but we do know that:
+                            //
+                            //  - if i == 0, we have no state machine, and must from scratch
+                            //  - if i == skiplist.len(), we scan from the last state machine
+                            //  - otherwise, the state machine from i-1 eventually encounters addr
+                            if i == 0 {
+                                unit.line_rows(&self.debug_line)
+                                    .chain_err(|| "cannot get line rows for unit")?
+                            } else {
+                                current = Some(skiplist[i - 1].2);
+                                rowi = (i + 1) * self.cache_every;
+                                skiplist[i - 1].1.clone()
+                            }
+                        }
+                    }
+                } else {
+                    // the skiplist was poisoned -- fall back to linear scan
+                    unit.line_rows(&self.debug_line)
+                        .chain_err(|| "cannot get line rows for unit")?
+                }
+            };
 
             // Now, find the last row before a row with a higher address than the one we seek.
-            let mut current = None;
-            while let Ok(Some((_, row))) = rows.next_row() {
-                if row.address() <= addr {
-                    if row.end_sequence() {
-                        current = None;
-                    } else {
-                        // Might be the right row, but we won't know until we see the next one.
-                        // The .clone is needed so we can keep iterating
-                        current = Some(*row);
+            let mut praddr = 0;
+            let mut skipseq = false;
+            while let Ok(Some((_, &row))) = rows.next_row() {
+                if row.end_sequence() {
+                    current = None;
+                    skipseq = false;
+                    continue;
+                }
+
+                if skipseq {
+                    continue;
+                }
+
+                let raddr = row.address();
+                if raddr == 0 {
+                    // NOTE: It is unclear if these rows are valid.
```

Comment:
Hmm, the question is if a single zero entry should invalidate the entire sequence... I don't know why we should require this? Without `skipseq = true` here, we'll simply move to the next entry, and if that is still lower, then we'll skip the sequence...